carbon-design-system/carbon-addons-iot-react
# fix(datetimepicker): pin flatpickr to fix absolute date range

Closes #2592 

**Summary**

Since Carbon 10.30/7.30 there was an upgrade to flatpickr which bumped it from 4.6.1 to 4.6.9. Since then, absolute DateTimePicker has been broken in our library. It will not allow you to select a new DateTimeRange when clicking days in the calendar. There is a work-around. If you follow the instructions below it will allow you to set a date. However, after much trying to programmatically change focus and/or blur the calendar I was unable to come up with a work-able solution. Our current implementation using the useEffect to set which input is currently focused causes this new version of flatpickr to fire the `onChange` event twice--first with the correct new dates--then again with the original dates thus overwriting the new dates and therefore appears as if no new date is set. I attempted to take out this focus code in the useEffect, but then, we're stuck with only the "end" date being change while the "start" date always stays 04/06/2020 (in the storybook example). 

Edit: 

I think I've narrowed the error down to this change: https://github.com/flatpickr/flatpickr/blame/ac9b752381102118cf4f44b5353fe30811b56209/src/index.ts#L1612 in flatpickr. That relatedTarget check causes the setDate _not_ to fire, so the start date returned to us in the onChange handler is always the same.

Previous functionality is available in this deploy preview https://deploy-preview-2088--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-datetime-picker--selected-absolute

#### Work-around
1. Open the DateTimePicker
2. Click a new day in the calendar.
3. Click away from the calendar (just gray space in the body if using the storybook
4. Click the new day you want to set

**Change List (commits, features, bugs, etc)**

- pinned flatpickr to version 4.6.1 (matching the same version used by Angular as well) -- I was unable to get it to resolve correctly when attempting to do it only for the react project.

**Acceptance Test (how to verify the PR)**

- Open the DateTimePicker "Selected absolute" story
- Click the DateTimePicker to open it.
- Click "April 8, 2020"
- The Date range should now change to days "6->8" being highlighted.
- Click "April 15, 2020"
- The Date range should now change to days "8->15" being highlighted.
- Click "April 2, 2020"
- The Date range should now change to days "2->15" being highlighted.

**Regression Test (how to make sure this PR doesn't break old functionality)**

- other features work as previously
